File: source/modulo_core/src/translators/message_readers.cpp
#include "modulo_core/translators/message_readers.hpp"

namespace modulo_core::translators {

static Eigen::Vector3d read_point(const geometry_msgs::msg::Point& message) {
  return {message.x, message.y, message.z};
}

static Eigen::Vector3d read_vector3(const geometry_msgs::msg::Vector3& message) {
  return {message.x, message.y, message.z};
}

static Eigen::Quaterniond read_quaternion(const geometry_msgs::msg::Quaternion& message) {
  return {message.w, message.x, message.y, message.z};
}

void read_message(state_representation::CartesianState& state, const geometry_msgs::msg::Accel& message) {
  state.set_linear_acceleration(read_vector3(message.linear));
  state.set_angular_acceleration(read_vector3(message.angular));
}

void read_message(state_representation::CartesianState& state, const geometry_msgs::msg::AccelStamped& message) {
  state.set_reference_frame(message.header.frame_id);
  read_message(state, message.accel);
}

void read_message(state_representation::CartesianState& state, const geometry_msgs::msg::Pose& message) {
  state.set_position(read_point(message.position));
  state.set_orientation(read_quaternion(message.orientation));
}

void read_message(state_representation::CartesianState& state, const geometry_msgs::msg::PoseStamped& message) {
  state.set_reference_frame(message.header.frame_id);
  read_message(state, message.pose);
}

void read_message(state_representation::CartesianState& state, const geometry_msgs::msg::Transform& message) {
  state.set_position(read_vector3(message.translation));
  state.set_orientation(read_quaternion(message.rotation));
}

void read_message(state_representation::CartesianState& state, const geometry_msgs::msg::TransformStamped& message) {
  state.set_reference_frame(message.header.frame_id);
  state.set_name(message.child_frame_id);
  read_message(state, message.transform);
}

void read_message(state_representation::CartesianState& state, const geometry_msgs::msg::Twist& message) {
  state.set_linear_velocity(read_vector3(message.linear));
  state.set_angular_velocity(read_vector3(message.angular));
}

void read_message(state_representation::CartesianState& state, const geometry_msgs::msg::TwistStamped& message) {
  state.set_reference_frame(message.header.frame_id);
  read_message(state, message.twist);
}

void read_message(state_representation::CartesianState& state, const geometry_msgs::msg::Wrench& message) {
  state.set_force(read_vector3(message.force));
  state.set_torque(read_vector3(message.torque));
}

void read_message(state_representation::CartesianState& state, const geometry_msgs::msg::WrenchStamped& message) {
  state.set_reference_frame(message.header.frame_id);
  read_message(state, message.wrench);
}

void read_message(state_representation::JointState& state, const sensor_msgs::msg::JointState& message) {
  state.set_names(message.name);
  if (!message.position.empty()) {
    state.set_positions(Eigen::VectorXd::Map(message.position.data(), message.position.size()));
  }
  if (!message.velocity.empty()) {
    state.set_velocities(Eigen::VectorXd::Map(message.velocity.data(), message.velocity.size()));
  }
  if (!message.effort.empty()) {
    state.set_torques(Eigen::VectorXd::Map(message.effort.data(), message.effort.size()));
  }
}

void read_message(bool& state, const std_msgs::msg::Bool& message) {
  state = message.data;
}

void read_message(double& state, const std_msgs::msg::Float64& message) {
  state = message.data;
}

void read_message(std::vector<double>& state, const std_msgs::msg::Float64MultiArray& message) {
  state = message.data;
}

void read_message(int& state, const std_msgs::msg::Int32& message) {
  state = message.data;
}

void read_message(std::string& state, const std_msgs::msg::String& message) {
  state = message.data;
}
}// namespace modulo_core::translators

File: source/modulo_core/modulo_core/translators/message_readers.py
from typing import List, TypeVar, Union

import clproto
import geometry_msgs.msg as geometry
import state_representation as sr
from modulo_core import EncodedState
from modulo_core.exceptions import MessageTranslationError
from sensor_msgs.msg import JointState

DataT = TypeVar('DataT')
MsgT = TypeVar('MsgT')
StateT = TypeVar('StateT')


def read_xyz(message: Union[geometry.Point, geometry.Vector3]) -> List[float]:
    """
    Helper function to read a list from a Point or Vector3 message.

    :param message: The message to read from
    :return: The vector coefficients as a list
    """
    return [message.x, message.y, message.z]


def read_quaternion(message: geometry.Quaternion) -> List[float]:
    """
    Helper function to read a list of quaternion coefficients (w,x,y,z) from a Quaternion message.

    :param message: The message to read from
    :return: The quaternion coefficients as a list
    """
    return [message.w, message.x, message.y, message.z]


def read_message(state: StateT, message: MsgT) -> StateT:
    """
    Convert a ROS message to a state_representation State type.

    :param state: The state to populate
    :param message: The ROS message to read from
    :raises MessageTranslationError if the message could not be read
    :return: The populated state
    """
    if not isinstance(state, sr.State):
        raise MessageTranslationError("This state type is not supported.")
    if isinstance(state, sr.CartesianState):
        if isinstance(message, geometry.Accel):
            state.set_linear_acceleration(read_xyz(message.linear))
            state.set_angular_acceleration(read_xyz(message.angular))
        elif isinstance(message, geometry.Pose):
            state.set_position(read_xyz(message.position))
            state.set_orientation(read_quaternion(message.orientation))
        elif isinstance(message, geometry.Transform):
            state.set_position(read_xyz(message.translation))
            state.set_orientation(read_quaternion(message.rotation))
        elif isinstance(message, geometry.Twist):
            state.set_linear_velocity(read_xyz(message.linear))
            state.set_angular_velocity(read_xyz(message.angular))
        elif isinstance(message, geometry.Wrench):
            state.set_force(read_xyz(message.force))
            state.set_torque(read_xyz(message.torque))
        else:
            raise MessageTranslationError("The provided combination of state type and message type is not supported")
    elif isinstance(message, JointState) and isinstance(state, sr.JointState):
        try:
            state.set_names(message.name)
            if len(message.position):
                state.set_positions(message.position)
            if len(message.velocity):
                state.set_velocities(message.velocity)
            if len(message.effort):
                state.set_torques(message.effort)
        except Exception as e:
            raise MessageTranslationError(f"{e}")
    else:
        raise MessageTranslationError("The provided combination of state type and message type is not supported")
    return state


def read_stamped_message(state: StateT, message: MsgT) -> StateT:
    """
    Convert a stamped ROS message to a state_representation State type.

    :param state: The state to populate
    :param message: The ROS message to read from
    :raises MessageTranslationError if the message could not be read
    :return: The populated state
    """
    if isinstance(message, geometry.AccelStamped):
        read_message(state, message.accel)
    elif isinstance(message, geometry.PoseStamped):
        read_message(state, message.pose)
    elif isinstance(message, geometry.TransformStamped):
        read_message(state, message.transform)
        state.set_name(message.child_frame_id)
    elif isinstance(message, geometry.TwistStamped):
        read_message(state, message.twist)
    elif isinstance(message, geometry.WrenchStamped):
        read_message(state, message.wrench)
    else:
        raise MessageTranslationError("The provided combination of state type and message type is not supported")
    state.set_reference_frame(message.header.frame_id)
    return state


def read_std_message(message: MsgT) -> DataT:
    """
    Read the data field of a std_msg.msg message.

    :param message: The message to read from
    :return: The data field of the message
    """
    return message.data


def read_clproto_message(message: EncodedState) -> StateT:
    """
    Convert an EncodedState message to a state_representation State type.

    :param message: The EncodedState message to read from
    :raises MessageTranslationError if the message could not be read:
    :return: The decoded clproto message
    """
    try:
        return clproto.decode(message.data.tobytes())
    except Exception as e:
        MessageTranslationError(f"{e}")
